File: src/create_model.py
from dslListener import dslListener
from antlr4 import *
from utils import destringify
from model import Model,Component
from phys import Dimension,Point
from datasheets import process_datasheet_prop
from known_packages import findKnownPackage

COMPONENT_INDEX_STRING=""

"""
    access_suffix: ('.' ID index?)
"""
def constant_fold_field(comp, access_suffixes):
    name0 = str(access_suffixes[0].ID())
    name1 = str(access_suffixes[1].ID())
    
    #print("constant fold in component " + str(names) + ", list = " + str(self.table_list))
    table = comp.find_table(name0)
    row = table.find_row_by_key(name1)
    value = int(row.get(1).string)
    print("PDF TABLE LOOKUP["+name0+"."+name1+"] = " + str(value))
    return value


"""
access: ID index? access_suffix*
access_suffix: ('.' ID index?)
"""
def constant_fold_access(model, access):
    name  = str(access.ID())
    if len(access.access_suffix()) == 0:
        if access.index() == None:
            return model.constants[name]
        
    comp = model.find_component(name)
    if comp != None:
        return constant_fold_field(comp, access.access_suffix())
    unimpl()


def constant_fold_primary(model, p):
    if p.length != None:
        name = str(p.ID())
        return model.current_component.resolve_length(name)
        
    if p.expr() != None:
        return constant_fold_expr(p.expr())
    if p.access() != None:
        return constant_fold_access(model, p.access())
    if p.unit != None:
        return Dimension(p.n, p.unit())
    print("unrecognized primary expr: " + str(p))
    unimpl()

def constant_fold_expr(model, expr):
    p = expr.primary()
    left = constant_fold_primary(model, p[0])
        
    if len(p) == 1:
        return left
    else:
        right = constant_fold_primary(model, p[1])        
        operand = expr.op().operand.text;
        return right.constant_fold(operand, left)


class ModelVar:
    def __init__(self, name, value):
        self.name = name
        self.value = value


class ModelContext:
    def __init__(self):
        self.vars = []

    def add(self, var):
        self.vars.append(var)

    def indexed_comp_name(self, name, index):
        for v in self.vars:
            if v.name == str(index.ID()):
                return name + COMPONENT_INDEX_STRING + str(v.value)            
        not_found()

    def indexed_pin_name(self, name, index):
        for v in self.vars:
            #print("EXAMINE " + v.name + " vs " + str(index.ID()))
            if v.name == str(index.ID()):
                return name + "" + str(v.value)
        print("ERRROR: failed to find " + str(index.ID()))
        not_found()

# access: ID index? access_suffix*
# 
# comp.P
# comp.P[x]
# comp[x].P
def access_to_component_pin(model, access, context, odd):
    name = str(access.ID())
    if access.index() != None:
        name = context.indexed_comp_name(name, access.index())

    comp = model.find_component(name)
    return comp.get_pin_by_suffixes(access.access_suffix(), context, odd)


def process_location(comp, loc_prop_list):
    model = comp.model
    for loc in loc_prop_list:
        # AT:
        model.current_component = comp
        sx = constant_fold_expr(model, loc.expr()[0])
        sy = constant_fold_expr(model, loc.expr()[1])

        comp.current_position = comp.fixed_position = Point(sx, sy, 0)        
        comp.transpose(None, comp.fixed_position, None, None)
        
        
def process_dimensions(comp, dim_prop_list):
    model = comp.model
    model.current_component = comp
    for dim in dim_prop_list:
        if dim.width != None:
            comp.width = constant_fold_expr(model, dim.width)
        if dim.height != None:
            comp.height = constant_fold_expr(model, dim.height)
        if dim.layers != None:
            comp.layers = constant_fold_expr(model, dim.layers)

    if not comp.has_data_sheet:
        sx = Dimension(0, "cm")
        sy = Dimension(0, "cm")
        if comp.component_type != None:
            pkg = findKnownPackage(comp.component_type)
            pkg.create_outline(comp)
        else:
            print("COMPONENT "+ comp.name + " has no assigned type yet")
            comp.outline.addRect(Point(sx, sy, comp.layers),
                                 Point(comp.width, comp.height, comp.layers))



def add_dimensions(comp, ctxt):
    if ctxt.component_property() != None:
        for p in ctxt.component_property():                    
            process_dimensions(comp, p.dim_prop())

            
def add_location(comp, ctxt):
    if ctxt.component_property() != None:
        for p in ctxt.component_property():                    
            process_location(comp, p.location_prop())
            
def add_datasheet_props(comp, ctxt):
    if ctxt.component_property() != None:
        for p in ctxt.component_property():                    
            process_datasheet_prop(comp, p.datasheet_prop())

def preprocess_component(model, comp, props):
    if comp.name == "board":
        comp.fixed_position = Point(Dimension(0, "mm"),
                                    Dimension(0, "mm"),
                                    0)
    
    model.current_component = comp
    for p in props:
        if len(p.datasheet_prop()) > 0:
            comp.has_data_sheet = True
        if p.component_type != None:
            comp.component_type = destringify(p.component_type.text)
            #print("SAW COMPONENT TYPE: "+ comp.component_type)
            
    for p in props:
        if p.pin_name() != None:
            for pin_name in p.pin_name().ID():
                pin = comp.add_pin(str(pin_name))
                for k in p.pin_prop():
                    pin.mode = k.pinmode
                        


            
class ModelListener(dslListener):
    def __init__(self, model):
        self.model = model

    def enterConstant(self, ctxt):
        #constant: 'const' ID '=' expr ';';
        name = str(ctxt.ID())
        expr = ctxt.expr()
        print("resolve constant: " + name)
        self.model.constants[name] = constant_fold_expr(self.model, expr)

    def add_connections(self, mctxt, ctxt):
        for conn in ctxt.connection():
            for k in range(0, len(conn.access())-1):
                from_access = conn.access()[k]
                to_access   = conn.access()[k + 1]
                from_pin   = access_to_component_pin(self.model, from_access, mctxt, True)
                to_pin     = access_to_component_pin(self.model, to_access, mctxt, False)
                from_pin.add_connection(to_pin)            

    def enterNetwork(self, ctxt):
        names = ctxt.object_name().ID()        
        if len(names) == 1:            
            mctxt = ModelContext()
            self.add_connections(mctxt, ctxt)
        else:
            limit = str(names[2])
            count = self.model.constants[limit]
            for i in range(0, count):
                var = ModelVar(str(names[1]), i)
                mctxt = ModelContext()
                mctxt.add(var)
                self.add_connections(mctxt, ctxt)

    
    def enterComponent(self, ctxt):
        names = ctxt.object_name().ID()
        #print("EXAMINE COMPONENT: " + str(names))
        if len(names) == 1:
            comp = Component(self.model, str(names[0]), False)
            preprocess_component(self.model, comp, ctxt.component_property())
            self.model.components.append(comp)

            add_dimensions(comp, ctxt)
            add_datasheet_props(comp, ctxt)
            add_location(comp, ctxt)            
        else:
            limit = str(names[2])
            count = self.model.constants[limit]
            for i in range(0, count):
                comp = Component(self.model, str(names[0]) + COMPONENT_INDEX_STRING + str(i), False)
                preprocess_component(self.model, comp, ctxt.component_property())
                self.model.components.append(comp)

                if ctxt.component_property() != None:
                    for p in ctxt.component_property():
                        process_dimensions(comp, p.dim_prop())
                
            
def read_model(tree):
    model = Model()
    ds = ModelListener(model)
    walker = ParseTreeWalker()
    walker.walk(ds, tree)
    return model

File: src/create_model.cc
#include "create_model.h"
#include <stdlib.h>

static constexpr double BOUNDING_BOX_MARGIN_MM = 0.5;

MillimeterPoint::MillimeterPoint(const Point &p)
  : x(((double)p.x) / Point::POINT_PRECISION),
    y(((double)p.y) / Point::POINT_PRECISION),
    layer(p.layer)
{
}



static
int constant_fold_expr(Model *model,
		       Component *current_component,
		       dslParser::ExprContext *expr);

static
int constant_fold_field(Component *comp,
			const std::vector<dslParser::Access_suffixContext *>& access_suffixes)
{
  auto name0 = access_suffixes[0]->ID()->getText();
  auto name1 = access_suffixes[1]->ID()->getText();
    
  //#print("constant fold in component " + str(names) + ", list = " + str(self.table_list));
  auto table = comp->find_table(name0);
  auto row = table->find_row_by_key(name1);
  if (! row)
    {
      fprintf(stderr, "failed to find row '%s' of table '%s'\n",
	      name1.c_str(), name0.c_str());
      table->dump();
      abort();
    }
  auto value = atoi(row->get(1)->string.c_str());
  printf("PDF TABLE LOOKUP[%s.%s] = %d\n", name0.c_str(), name1.c_str(), value);
  return value;
}

static
int constant_fold_access(Model *model, dslParser::AccessContext *access)
{
  auto name = access->ID()->getText();
  if (access->access_suffix().size() == 0)
    {
      if (access->index() == NULL)
	{
	  return model->info->constants[name];
	}
    }
  
  auto comp = model->find_component(name);
  if (comp)
    {
      return constant_fold_field(comp, access->access_suffix());
    }
  abort();
}


static
int constant_fold_primary(Model *model,
			  Component *current_component,
			  dslParser::PrimaryContext *p)
{
  if (p->length)
    {
      auto name = p->ID()->getText();
      return current_component->resolve_length(name);
    }
  if (p->expr())
    {
      return constant_fold_expr(model, current_component, p->expr());
    }
  if (p->access())
    {
      return constant_fold_access(model, p->access());
    }
  auto unit = p->unit()->name->getText();
  auto value = atof(p->n->getText().c_str());
  if (unit == "cm")
    return value * 10;
  if (unit == "mm")
    return value;
  if (unit == "layer")
    return value;
  abort();
}

static
int constant_fold_expr(Model *model,
		       Component *current_component,
		       dslParser::ExprContext *expr)
{
  auto p = expr->primary();
  auto left = constant_fold_primary(model, current_component, p[0]);
  if (p.size() == 1)
    {
      return left;
    }
  else
    {
      auto right = constant_fold_primary(model, current_component, p[1]);
      auto operand = expr->op()->operand->getText();
      if (operand == "+")
	return left + right;
      if (operand == "*")
	return left * right;
      if (operand == "/")
	return left / right;
      if (operand == "%")
	return left % right;
      abort();
    }
  return 0;
}

   
Pin *Component::get_pin_by_name(const std::string &s)
{
  const unsigned count = pins.size();
  for (unsigned i=0;i<count;i++)
    {
      auto p = pins[i];
      
      if (p->info->name == s)
	{
	  return p;   
	}
    }
  fprintf(stderr, "failed to find pin %s\n", s.c_str());
  abort();
}

Pin *Component::get_pin_by_suffixes(const std::vector<dslParser::Access_suffixContext*> &suffixes,
				    ModelContext &context,
				    bool odd)
{
  if (suffixes.size() == 0)
    {
      if (pins.size() == 1)
	{
	  return pins[0];
	}
      
      if (pins.size() > 2)
	{
	  printf("don't know which pin to address, please make it explicit: %s\n", this->info->name.c_str());
	  abort();
	}    
      return pins[odd];
    }
  else
    {
      auto s0 = suffixes[0];
      auto name = s0->ID()->getText();
      if (s0->index() == 0)
	{
	  return get_pin_by_name(name);
	}
      else
	{
	  return get_pin_by_name(context.indexed_pin_name(name,
							  s0->index()));
	}
    }
}


Pin *access_to_component_pin(Model *model,
			     dslParser::AccessContext* access,
			     ModelContext &context,
			     bool odd)
{
  auto name = access->ID()->getText();
  if (access->index())
    {
      name = context.indexed_comp_name(name, access->index());
    }
  
  auto comp = model->find_component(name);
  return comp->get_pin_by_suffixes(access->access_suffix(),
				   context,
				   odd);
}

void preprocess_component(Model *model,
			  Component *comp,
			  const std::vector<dslParser::Component_propertyContext *>& props)
{
  if (comp->info->name == "board")
    {
      comp->info->fixed_position = new Point;
    }
  
  //model->current_component = comp;
  
  for (auto p : props)
    {
      if (p->datasheet_prop().size() > 0)
	{
	  comp->info->has_data_sheet = true;
	}
      if (p->component_type)
	{
	  comp->info->component_type = destringify(p->component_type->getText());
	  //fprintf(stderr, "SAW COMPONENT TYPE: '%s'\n", comp->component_type.c_str());
	}
    }
  
  for (auto p : props)
    {
      if (p->pin_name())
	{
	  for (auto pin_name : p->pin_name()->ID())
	    {
	      std::string normalized_pin_name = normalize_ident(pin_name->getText());
	      //fprintf(stderr, "normalized to %s\n", normalized_pin_name.c_str());

	      auto info = new PinInfo(normalized_pin_name);
	      auto pin = comp->add_pin(info);
	      for (auto k : p->pin_prop())
		{
		  auto m = k->pinmode->getText();

		  if (m == "input")
		    pin->info->mode = PinInfo::Mode::INPUT;
		  else if (m == "output")
		    pin->info->mode = PinInfo::Mode::OUTPUT;
		  else if (m == "inout")
		    pin->info->mode = PinInfo::Mode::INOUT;
		  else if (m == "analog_ground")
		    pin->info->mode = PinInfo::Mode::ANALOG_GROUND;
		  else if (m == "digital_ground")
		    pin->info->mode = PinInfo::Mode::DIGITAL_GROUND;
		  else {
		    fprintf(stderr, "unrecognized mode: %s\n", m.c_str());
		    abort();
		  }
		}
	    }
	}    
    }
}

static
void add_datasheet_props(Component *comp, dslParser::ComponentContext *ctxt)
{
  for (auto p : ctxt->component_property())
    {
      process_datasheet_prop(comp, p->datasheet_prop());
    }
}


static
void process_location(Component *comp,
		      dslParser::Location_propContext * loc)
{
  if (!loc)
    {
      return;
    }
  
  auto model = comp->model;

  if (loc->rotatable)
    {
      comp->info->is_rotateable = true;
    }
  
  if (loc->x)
    {
      //model->current_component = comp;
      auto sx = constant_fold_expr(model, comp, loc->x);
      auto sy = constant_fold_expr(model, comp, loc->y);

      int layer = 0;
      if (loc->layer)
	{
	  layer = atoi(loc->layer->getText().c_str());
	}
	  
      // constant fold already changed every position to millimeters:
      comp->info->fixed_position = new Point(MillimeterPoint(sx,
							     sy,
							     layer));
      comp->transpose(*comp->info->fixed_position);
      comp->set_layer(layer);
    }
}


static
void add_location(Component *comp, dslParser::ComponentContext *ctxt)
{
  for (auto p : ctxt->component_property())
    {
      process_location(comp, p->location_prop());
    }
}

static
void do_process_dimensions(Component *comp,
			   const std::vector<dslParser::Dim_propContext *> &dim_prop_list)
{
  assert(dim_prop_list.size() > 0);
  auto model = comp->model;
  //model->current_component = comp;
  
  for (auto dim : dim_prop_list)
    {
      if (dim->width)
	{
	  comp->info->dim.x = constant_fold_expr(model, comp, dim->width);
	}
      
      if (dim->height)
	{
	  comp->info->dim.y = constant_fold_expr(model, comp, dim->height);
	}
      
      if (dim->layers)
	{
	  comp->info->dim.layer = constant_fold_expr(model, comp, dim->layers);
	}
    }
    
  if (comp->info->name == "board")
    {
      model->info->board_dim = comp->info->dim;
    }
}


static
void add_dimensions(Component *comp, dslParser::ComponentContext *ctxt)
{
  for (auto p : ctxt->component_property())
    {
      if (p->dim_prop().size() > 0)
	{
	  do_process_dimensions(comp, p->dim_prop());
	}
    }
}

static
void process_component_type(Model *model,
			    Component *comp)
{  
  if (! comp->info->has_data_sheet)
    {
      if (comp->info->component_type != "")
	{
	  auto pkg = findKnownPackage(comp->info->component_type);
	  pkg->create_outline(comp);
	}
      else
	{
	  if (! comp->info->is_board)
	    {
	      printf("COMPONENT '%s' has no assigned type yet\n", comp->info->name.c_str());
	    }
	  comp->outline.addRect(Point(MillimeterPoint(0,
						      0,
						      comp->info->dim.layer)),
				Point(comp->info->dim));
	}
    }
  else
    {
      fprintf(stderr, "component %s has a datasheet!\n", comp->info->name.c_str());
    }
}


void ModelCreatorListener::add_connections(ModelContext &mctxt,
					   dslParser::NetworkContext *ctxt)
{
  for (auto conn : ctxt->connection())
    {
      for (unsigned k = 0; k < conn->access().size()-1; k++)
	{
	  auto from_access = conn->access()[k];
	  auto to_access   = conn->access()[k + 1];
	  auto from_pin   = access_to_component_pin(this->model,
						    from_access,
						    mctxt,
						    true); // odd
	  auto to_pin     = access_to_component_pin(this->model,
						    to_access,
						    mctxt,
						    false); // even

	  if (from_pin->get_layer() != to_pin->get_layer())
	    {
	      // connection of something on one layer to something
	      // on another layer. We hence must add a router.	      
	      Component *router = this->model->create_router(from_pin->center());
	      from_pin->add_connection(router->pins[0]);
	      router->pins[0]->set_layer(from_pin->get_layer());
	      
	      router->pins[1]->set_layer(to_pin->get_layer());
	      router->pins[1]->add_connection(to_pin);

	      assert(from_pin->get_layer() == router->pins[0]->get_layer());
	      assert(to_pin->get_layer()   == router->pins[1]->get_layer());	    
	    }
	  else
	    {
	      assert(from_pin->get_layer() == to_pin->get_layer());
	      from_pin->add_connection(to_pin);
	    }
	}
    }
}


void ModelCreatorListener::enterConstant(dslParser::ConstantContext *ctxt)
{
  //constant: 'const' ID '=' expr ';';
  auto name = ctxt->ID()->getText();
  auto expr = ctxt->expr();
  //std::cout << "resolve constant: " << name;
  assert(current_component);
  this->model->info->constants[name] = constant_fold_expr(this->model,
							  current_component,
							  expr);
}


void ModelCreatorListener::enterNetwork(dslParser::NetworkContext *ctxt)
{
  auto names = ctxt->object_name()->ID();
  if (names.size() == 1)
    {
      ModelContext mctxt;
      this->add_connections(mctxt, ctxt);
    }
  else
    {
      auto limit = names[2]->getText();
      auto count = this->model->info->constants[limit];
      for (int i = 0; i < count; i++) {
	auto var = ModelVar(names[1]->getText(), i);
	ModelContext mctxt;
	mctxt.add(var);
	this->add_connections(mctxt, ctxt);
      }
    }
}


void Component::add_bounding_box()
{
  min_max_t d;
  outline.minmax(d);
  for (auto p : pins)
    {
      p->outline.minmax(d);
    }

  MillimeterLength ml(BOUNDING_BOX_MARGIN_MM);

  d.min = d.min.sub(MillimeterPoint(ml, ml, d.min.layer));
  d.max = d.max.add(MillimeterPoint(ml, ml, d.min.layer));
  bounding_box.addRect(d.min, d.max);
  bounding_box.set_layer(outline.get_layer());
}


void ModelCreatorListener::enterComponent(dslParser::ComponentContext *ctxt)
{
  auto names = ctxt->object_name()->ID();
  //utils::print("EXAMINE COMPONENT: ", utils::str(names));
  if (names.size() == 1)
    {
      ComponentInfo *info = new ComponentInfo(names[0]->getText(), false);
      
      auto comp = new Component(info, this->model);
      this->current_component = comp;
      preprocess_component(this->model, comp, ctxt->component_property());
      this->model->components.push_back(comp);
	
      add_dimensions(comp, ctxt);
      process_component_type(model, 
			     comp);
      add_datasheet_props(comp, ctxt);
      add_location(comp, ctxt);
      comp->add_bounding_box();
    }
  else
    {
      auto limit = names[2]->getText();
      auto count = this->model->info->constants[limit];
      for (int i = 0; i< count; i++)
	{
	  auto info = new ComponentInfo(names[0]->getText() + str(i), false);
	  
	  auto comp = new Component(info, this->model);
	  this->current_component = comp;	    
	  preprocess_component(this->model, comp, ctxt->component_property());
	  this->model->components.push_back(comp);
	    
	  for (auto p : ctxt->component_property())
	    {
	      if (p->dim_prop().size() > 0)
		{
		  do_process_dimensions(comp, p->dim_prop());
		}	      
	    }
	  process_component_type(model,
				 comp);
	  comp->add_bounding_box();
	}
    }
}

